<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [16]:
def fibonacci(n):
    if n==1 or n==2:
        return 1
    else: 
        return fibonacci(n-1) + fibonacci(n-2)

In [26]:
print(fibonacci(1))
print(fibonacci(2))
print(fibonacci(3))
print(fibonacci(4))
print(fibonacci(5))
print(fibonacci(6))

1
1
2
3
5
8


#### Q 1.1.2
What's the complexity of your implementation?

**Answer**: This implementation is $O(2^n)$ in execution time, since every recursive step has to recompute all the previous fibonacci tree of values. 

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [44]:
def fibonacci2(n):
    def fib_helper(n,curr,i,acc):
        if n!=i:
            if i==1:
                return fib_helper(n,1,i+1,1)
            else:
                return fib_helper(n,acc,i+1,acc+curr)
        return acc
    return fib_helper(n,1,1,1)

In [50]:
print(fibonacci2(1))
print(fibonacci2(2))
print(fibonacci2(3))
print(fibonacci2(4))
print(fibonacci2(5))
print(fibonacci2(6))
print(fibonacci2(100))

1
1
2
3
5
8
354224848179261915075


#### Q 1.1.4
What's the complexity of your implementation?

**Answer**: This implementation is $O(n)$, since the tail recursive function only has to do one pass through the data. The acc and curr values memoize the values needed to compute the next fibonacci values efficiently.

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


**Answer**: If the Fibonacci values are going to be frequently queried, it might be useful to store them in a dictionary. It could also be implemented using loops and two auxiliary values instead of recursion if the language does not support tail recursion (will deplete stack memory instead of reusing it).

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

**Answer:** This function uses Monte Carlo sampling to approximate the value of $\pi$. The idea is take a square plotted in the range $[0,1]$ and generate random points within that square. Imagine that there is a circle within the square with radius 0.5. The area of the circle is then $A_c = \pi*0.5^2 = \frac{\pi}{4}$. Given enough random points generated, the probability that a point will be generated within the circle is equal to its area. We can leverage this to get an approximation of $\pi$: just calculate $\pi = 4*|inCircle|/|totalSamples|$.

In [73]:
import random
def compute_pi(samples):
    acc_in_circle=0
    for i in range(samples):
        x = random.random()
        y = random.random()
        if x**2+y**2<=1:
            acc_in_circle+=1
    return 4*float(acc_in_circle)/samples

In [74]:
compute_pi(50000000)

3.14175472

### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


**Answer:** 

The no free lunch theorem states that no approximation algorithm can be superior to all others, when averaged across all possible datasets. Said in another way, even though KNN is a universal approximator, we can put forth datasets where other methods outperform KNN. Therefore, it is still valuable to research different approximators that might achieve better generalizations in different contexts.

More practically, KNN has several drawbacks. Since it's non-paremetric, the memory requirements are equal to storing the whole dataset. Furthermore, even though it converges to the optimal solution as $N$ and $k$ increases, the complexity of the inference is also linear with respect to $k$. This computational limitations may make KNN undesirable in many practical settings, as compared to parametric models such as neural networks that may scale better in memory with large data and have a relatively low computational tax in inference time.

### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.

